### PR TITLE
Compress and extract functions

### PR DESCRIPTION
**Pull request by magalilemes**
This PR adds compression and extraction functions to kwlib.

### CODE REVIEW

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @magalilemes ,

First of all, thanks a lot for your PR.

I added a couple of comments.

Thanks

File: src/kwlib.sh [hunk 1]
```diff
@@ -387,3 +387,19 @@ function compress()
   cmd="tar -C $path_to_compress -czf $file_path ."
   cmd_manager "$flag" "$cmd"
 }
+
+# This function extracts a .tar.gz file to a given path
+#
+# @file_to_extract The path to the file to extract from
+# @path Where to extract the file
+# @flag How to display (or not) the command used
+function extract()
```
[the remark this answers is not in the document]
Comment (magalilemes) [reply src/kwlib.sh:396]: Maybe, but I feel like _extract_ is a more common word for this purpose.

File: tests/kwlib_test.sh [hunk 2]
```diff
@@ -392,4 +392,16 @@ function test_exit_msg()
...
+function test_compress()
+{
+  local path_to_compress='tests/samples/statistics'
+  local file_path="$SHUNIT_TMPDIR/compressed.tar.gz"
+  local output
+
+  output=$(compress "$path_to_compress" "$file_path" 'SUCCESS')
+  assertEquals "($LINENO)" "tar -C $path_to_compress -czf $file_path ." "$output"
+
+  assertTrue 'Compressed file was not created' "[[ -f $SHUNIT_TMPDIR/compressed.tar.gz ]]"
```
[the remark this answers is not in the document]
Comment (magalilemes) [reply tests/kwlib_test.sh:404]: This one test is doing that: it tries to compress the directory `tests/samples/statistics`, sees if the command used is correct and then checks if the compressed file was created

File: src/kwlib.sh [hunk 3]
```diff
@@ -371,3 +371,38 @@ function exit_msg()
...
+#
+# @path_to_compress The directory to compress
+# @file_path Where the compressed directory will be stored
+# @compression_type compression program used
+# @flag How to display (or not) the command used
+function generate_tarball()
+{
+  local path_to_compress="$1"
+  local file_path="$2"
+  local compression_type="$3"
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:385]: Can we define a default compression type here?

File: src/kwlib.sh [hunk 4]
```diff
@@ -371,3 +371,38 @@ function exit_msg()
...
+  local flag="$4"
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:386]: Let's define SILENT as the default flag.

File: src/kwlib.sh [hunk 5]
```diff
@@ -371,3 +371,38 @@ function exit_msg()
...
+  local ret
+  local cmd
+
+  if [[ ! -d "$path_to_compress" ]]; then
+    complain "$path_to_compress" 'does not exist'
+    exit 22 #EINVAL
+  fi
+
+  if [[ -n "$compression_type" ]]; then
+    compression_type="--$compression_type"
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:396]: Can you check if we can validate this compression flag? Maybe the `tar` command provides something; if not, we can try to compress a dummy file just for checking (I would prefer the first option).

File: src/kwlib.sh [hunk 6]
```diff
@@ -371,3 +371,38 @@ function exit_msg()
...
+  fi
+
+  cmd="tar -C $path_to_compress $compression_type -cf $file_path ."
+  cmd_manager "$flag" "$cmd"
+
+  ret=$?
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:402]: Double-quote

File: src/kwlib.sh [hunk 7]
```diff
@@ -371,3 +371,38 @@ function exit_msg()
...
+
+  if [[ "$ret" != 0 ]]; then
+    complain "Error archiving modules."
+    exit $ret
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:406]: Double-quote

File: tests/kwlib_test.sh [hunk 8]
```diff
@@ -392,4 +392,37 @@ function test_exit_msg()
...
+
+  output=$(tar -taf "$file_path" | sort -d)
+  while read -r f; do
+    if [[ ${expected_files[$count]} != "${f}" ]]; then
+      fail "$ID - Expected file \"${f}\" to be \"${expected_files[$count]}\""
+    fi
+    ((count++))
+  done <<< "$output"
```
Comment (rodrigosiqueira) [inline tests/kwlib_test.sh:422]: I think you can replace all of these codes by using `compare_command_sequence`.

File: src/kwlib.sh [hunk 9]
```diff
@@ -406,3 +406,35 @@ function generate_tarball()
...
+#
+# @file_to_extract The path to the file to extract from
+# @path Where to extract the file
+# @compression_type compression program used
+# @flag How to display (or not) the command used
+function extract_tarball()
+{
+  local file_to_extract="$1"
+  local path="$2"
+  local compression_type="$3"
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:420]: I'm not sure, but I think the option `-a` from tar command is an automatic mechanism to check the compression type and use it (see `-taf`). Let's use this option as a default.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
In general, lgtm. I added a couple of comments.

**Comment by rodrigosiqueira** [comment]
Hi @magalilemes , any progress here?

File: src/kwlib.sh [hunk 10]
```diff
@@ -1,5 +1,9 @@
 # NOTE: src/kw_config_loader.sh must be included before this file
 
+# Array with compression programs accepted by tar
+declare -ga compression_programs=('gzip' 'bzip2' 'lzip' 'lzma' 'lzop' 'zstd' 'xz'
+  'auto-compress')
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:5]: Can you confirm if we need to add a dependency in the `setup.sh` for some of these compression options? For example, I think lzma requires something extra.

File: src/kwlib.sh (hunk 3 above)
Comment (rodrigosiqueira) [reply src/kwlib.sh:436]: I think you forgot to set a default compression.

File: src/kwlib.sh (hunk 9 above)
Comment (rodrigosiqueira) [reply src/kwlib.sh:480]: Did you check the `-a` option? Can we use it as the default?

File: src/kwlib.sh [hunk 11]
```diff
@@ -418,3 +422,88 @@ function kw_parse_get_errors()
...
+
+  cmd="tar -C $path_to_compress $compression_type -cf $file_path ."
+  cmd_manager "$flag" "$cmd"
+
+  ret="$?"
+
+  if [[ "$ret" != 0 ]]; then
+    complain "Error archiving modules."
+    exit "$ret"
+  fi
```
Comment (rodrigosiqueira) [inline src/kwlib.sh:467]: It looks like that we are not testing this path. Could you add a test for that? Also, use single quotes.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Thanks @magalilemes ,
I just added a few comments.

File: src/kwlib.sh (hunk 9 above)
Comment (magalilemes) [reply src/kwlib.sh:480]: Yes, it's the same as `--auto-compress`, and I opted to set it as default a few lines below (line 457)

Comment (rodrigosiqueira) [reply src/kwlib.sh:480]: Nice, I missed that. Thanks.

File: src/kwlib.sh (hunk 10 above)
Comment (magalilemes) [reply src/kwlib.sh:5]: From what I searched, lzip and zstd don't come by default on Ubuntu, so users need to install them before using them with tar

**Comment by rodrigosiqueira** [comment]
Merged!
Thanks

**Pull request closed without merging**